### PR TITLE
add extra visual level for local view, view more only lose some detail.

### PR DESCRIPTION
**Pull request by lalawue**
add extra visual level to local image, we can zoom in even data source can not provide that level.

### CODE REVIEW

**Comment by dblock** [comment]
Thanks, this looks pretty cool. 

At the very lists it needs tests to be merged and an entry in CHANGELOG.

I wonder whether _extraVisualLevel_ is the best name for this. It's about _tileLevel_, so I think  _maxVisualTileLevel_ would be more consistent as it's also a max?

File: Classes/ARLocalTiledImageDataSource.m [hunk 1]
```diff
@@ -11,16 +11,51 @@
 
 @implementation ARLocalTiledImageDataSource
 
+- (id)init
+{
+    self = [super init];
+    if (self) {
```
Comment (dblock) [inline Classes/ARLocalTiledImageDataSource.m]: I don't think this should be initialized to anything, I would leave it `nil` and handle that.

File: Classes/ARLocalTiledImageDataSource.m [hunk 2]
```diff
@@ -11,16 +11,51 @@
...
+        _extralVisualLevel = 1;
+    }
+    return self;
+}
+
 - (UIImage *)tiledImageView:(ARTiledImageView *)imageView imageTileForLevel:(NSInteger)level x:(NSInteger)x y:(NSInteger)y
 {
-    NSString *filename = [NSString stringWithFormat:@"%@/%@/%@_%@.%@", self.tileBasePath, @(level), @(x), @(y), self.tileFormat];
-    return [UIImage imageWithContentsOfFile:filename];
+#define _PATH(_LEV, X, Y)   [NSString stringWithFormat:@"%@/%@/%@_%@.%@", self.tileBasePath, @(_LEV), @(X), @(Y), self.tileFormat];
```
Comment (dblock) [inline Classes/ARLocalTiledImageDataSource.m]: Just make a separate function for this if it's to be reused, defines are not local to a function if I am not mistaken, so this is actually global.

File: Classes/ARLocalTiledImageDataSource.m [hunk 3]
```diff
@@ -11,16 +11,51 @@
...
+    
+    if (level <= self.maxTileLevel) {
+        NSString *filename = _PATH(level, x, y);
+        return [UIImage imageWithContentsOfFile:filename];
+    }
+    
+    NSInteger gap = level - _maxTileLevel;
```
Comment (dblock) [inline Classes/ARLocalTiledImageDataSource.m]: Refactor this into `imageWithContentsOfFileWithGap` or something like that.

File: Classes/ARLocalTiledImageDataSource.m [hunk 4]
```diff
@@ -11,16 +11,51 @@
...
+    NSInteger glen = _tileSize >> gap;
+    NSInteger nm = 1<<gap;
+    NSInteger xoff=(x%nm)*glen, yoff=(y%nm)*glen;
```
Comment (dblock) [inline Classes/ARLocalTiledImageDataSource.m]: Please make this look a bit like the rest of the code, for example we do spaces around things like `<<` or `=`.

File: Classes/ARLocalTiledImageDataSource.m [hunk 5]
```diff
@@ -11,16 +11,51 @@
...
+    
+    CGImageRef imgRef = CGImageCreateWithImageInRect(oImg.CGImage, (CGRect){xoff, yoff, glen, glen});
+    oImg = [[UIImage alloc] initWithCGImage:imgRef];
+    CGImageRelease(imgRef);
+    
+    UIGraphicsBeginImageContextWithOptions(CGSizeMake(_tileSize, _tileSize), YES, 1.);
+    CGContextRef context = UIGraphicsGetCurrentContext();
+    CGContextSetInterpolationQuality(context, kCGInterpolationNone);
+    [oImg drawInRect:CGRectMake(0,0,_tileSize,_tileSize)];
+    UIImage *nImg = UIGraphicsGetImageFromCurrentImageContext();
```
Comment (dblock) [inline Classes/ARLocalTiledImageDataSource.m]: Minor, but why `oImg` and `nImg`? What's the `o` or `n` for? Maybe rename to full names such as `originalImage`, etc.?

File: Classes/ARLocalTiledImageDataSource.m [hunk 6]
```diff
@@ -38,7 +73,11 @@ - (NSInteger)minimumImageZoomLevelForImageView:(ARTiledImageView *)imageView
 
 - (NSInteger)maximumImageZoomLevelForImageView:(ARTiledImageView *)imageView
 {
-    return self.maxTileLevel;
+    return self.maxTileLevel + _extralVisualLevel;
```
Comment (dblock) [inline Classes/ARLocalTiledImageDataSource.m]: This means that the maximumZoomLevel has become +1 by default? I don't think the default behavior should change.

**Comment by dblock** [comment]
I made a bunch of comments. It has been a while since I touched this code, so forgive me if some questions are naive. Hang with me here and lets get this merged, it's a nice and useful function!

File: Classes/ARLocalTiledImageDataSource.m (hunk 6 above)
Comment (lalawue) [inline Classes/ARLocalTiledImageDataSource.m]: I have make a mistake, the _extralVisualLevel initial value is 0.

**Comment by lalawue** [comment]
Actually set maximumZoomScale of NATiledImageMapView can do the same thing, may be I should widthdraw the pull request?

**Comment by dblock** [comment]
So `maximumZoomScale` just works and accomplishes the same result? Then yes.

**Pull request closed without merging**